### ISSUE TITLE
Code Quality: Update Selectors in checks PPCv2.0

### ISSUE THREAD

**Issue by maxyinger** [GoogleForCreators/web-stories-wp#8289]
Porting over PPC to the new structure in PPCv2.0 we were just trying to get all the old code into the new structure. The old code took the entire story when seeing if a check was passing and failing. When we moved everything over to the new structure, we were trying to alter everything as little as possible because there were a lot of moving pieces and we didn't want to introduce more surface area for there to be regressions.

Because of the approach above, a lot of the checks are still pulling the entire story object:
```
const story = useStory();
const isCheckRendered = checkStoryForSomething(story);
```

However, with the new structure, we have better access to the context selector, so we can fine tune these `useStory()` uses to have a more fine tuned selector and only perform the check on the part of the story it needs vs the whole story object.

This should increase performance of the checks because they should only be re-rendering when parts of the story they're checking against update.

Completion of this task includes going through every check and making sure each check only re-renders and runs its check of the necessary parts of the story.

**Comment by dreamofabear** [comment]
WDYT about only giving checklist components the story `pages` but passed through `ChecklistProvider`? This would keep the context dependency local to the checklist.

**Comment by swissspidy** [comment]
Some checklist checks are for story metadata too, not only pages

**Comment by maxyinger** [comment]
> WDYT about only giving checklist components the story `pages` but passed through `ChecklistProvider`? This would keep the context dependency local to the checklist.

I guess I'm not completely following. I think having each check pull only what's needed from `useStory` should be the most performant result because it would result in the smallest subtrees getting diffed (only relevant checks and their children) when the story updates. 

I think these use-context-selector providers are just subscriptions held by traditional context providers under the hood, so maybe it wouldn't effect it at all perf wise if we passed any story context from `useStory` through `ChecklistProvider` . That being said, I think having the smallest chain between between the place the data is coming from and the component receiving it lends the most transparency with regard to perf.

**tldr** I'm not seeing the benefit, but open to it if there's some nice benefits I'm not quite grokking rn

**Comment by maxyinger** [comment]
@choumx Going through it a little more now, I think I might be getting at what you're saying more with regard to perf. Might see a better increase with elements with pageId being hoisted up to the `useStory` provider so the reduce only runs once every time the pages update, then we can just pull that into a check that runs on every element and only have to run a simple `.filter()` .

https://github.com/google/web-stories-wp/blob/9fce684d0a7aedba026b3b115c39e54233885869/assets/src/edit-story/components/checklist/utils/filterStoryElements.js#L29

Not sure if it'd be that much faster, but should be less flatMap calls and object instantiations every time `pages` updates (probably ~10x less). thoughts?

**Comment by dreamofabear** [comment]
Oh I was thinking more about context encapsulation rather than a strict performance benefit.

IMO it's cleaner to have a single `useStory` in the `ChecklistProvider` vs. a `useStory` in each checklist component since it keeps component dependencies more local. Though we'd also reduce duplication of code/work in the context selector as you mentioned.

**Comment by maxyinger** [comment]
> IMO it's cleaner to have a single useStory in the ChecklistProvider

Ah yea, completely agree but I think with this approach we were intentionally moving away from that so each check could have more fine grained logic to re-render based on the selector it passes to the `useStory` hook vs just running every check every time any part of the story updates.

I agree that I think the tradeoff is that it's not as clean, but I think the fact that we previously had perf issues with a more centralized setup warrants the current approach
